File: producer_consumer/producer_consumer.py
import threading
import time
import random

N             = 0
buffer        = []
total_items   = 0
in_ptr        = 0
out_ptr       = 0
produced_count = 0
consumed_count = 0

mutex = threading.Semaphore(1)
empty = None
full  = threading.Semaphore(0)
count_lock = threading.Lock()

done_producing = False  # flag: producers are done

def produce_item(producer_id):
    item = random.randint(1, 100)
    print(f"  [Producer-{producer_id}] Produced  item = {item}")
    time.sleep(0.5)
    return item

def insert_item(item):
    global in_ptr
    buffer[in_ptr] = item
    in_ptr = (in_ptr + 1) % N

def remove_item():
    global out_ptr
    item = buffer[out_ptr]
    out_ptr = (out_ptr + 1) % N
    return item

def consume_item(item, consumer_id):
    print(f"  [Consumer-{consumer_id}] Consumed  item = {item}")
    time.sleep(0.5)

def producer(producer_id):
    global produced_count, done_producing

    while True:
        # Reserve one production slot so total produced items never exceeds total_items.
        with count_lock:
            if produced_count >= total_items:
                break
            produced_count += 1

        item = produce_item(producer_id)

        # Acquire in this order: empty slot first, then mutex for buffer access.
        empty.acquire()    # wait(): wait if buffer is full
        mutex.acquire()    # wait(): enter critical section

        insert_item(item)
        print(f"  [Producer-{producer_id}] Inserted  item={item}  into slot={(in_ptr - 1) % N}")

        mutex.release()    # signal(): leave critical section
        full.release()     # signal(): notify consumer

    print(f"  [Producer-{producer_id}] Finished.")

def consumer(consumer_id):
    global consumed_count

    while True:
        # Timeout prevents permanent blocking when no more items will arrive.
        acquired = full.acquire(timeout=2)

        if not acquired:
            # Re-check shared counters under lock before deciding to stop.
            with count_lock:
                if consumed_count >= total_items:
                    break
            continue  # else keep waiting

        mutex.acquire()    # wait(): enter critical section

        if consumed_count >= total_items:
            mutex.release()
            # Wake another consumer that may still be waiting on full.
            full.release()
            break

        consumed_count += 1
        item = remove_item()
        print(f"  [Consumer-{consumer_id}] Removed   item={item}  from slot={(out_ptr - 1) % N}")

        mutex.release()    # signal(): leave critical section
        empty.release()    # signal(): notify producer

        consume_item(item, consumer_id)

    print(f"  [Consumer-{consumer_id}] Finished.")

def main():
    global N, buffer, total_items, empty

    print("╔══════════════════════════════════════════════════════╗")
    print("║  Producer-Consumer Problem  (Semaphore in Python)   ║")
    print("╚══════════════════════════════════════════════════════╝\n")

    N             = int(input("Enter buffer size           (N) : "))
    total_items   = int(input("Enter total items to produce    : "))
    num_producers = int(input("Enter number of producers       : "))
    num_consumers = int(input("Enter number of consumers       : "))

    # Pre-allocate fixed-size circular buffer and initialize free-slot count.
    buffer = [0] * N
    empty  = threading.Semaphore(N)

    print("\n─── Simulation Start ───────────────────────────────\n")

    producer_threads = [threading.Thread(target=producer, args=(i+1,)) for i in range(num_producers)]
    consumer_threads = [threading.Thread(target=consumer, args=(i+1,)) for i in range(num_consumers)]

    # Start all workers, then wait for a clean shutdown.
    for t in producer_threads: t.start()
    for t in consumer_threads: t.start()

    for t in producer_threads: t.join()
    for t in consumer_threads: t.join()

    print("\n─── Simulation End ──────────────────────────────────")
    print(f"  Total produced : {produced_count}")
    print(f"  Total consumed : {consumed_count}")
    print("╔══════════════════════════════════════════════════════╗")
    print("║                     All Done!                       ║")
    print("╚══════════════════════════════════════════════════════╝")

if __name__ == "__main__":
    main()

File: producer_consumer/producer_consumer.c
/*
 * Producer-Consumer Problem — Semaphore Solution in C
 */

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

// ─── Globals set by user ─────────────────────────────────────────────────────
int  N;             // buffer size
int *buffer;        // circular buffer array
int  in  = 0;       // next insert position
int  out = 0;       // next remove position
int  total_items;   // total items to produce

// ─── Semaphores ──────────────────────────────────────────────────────────────
sem_t mutex;   // binary semaphore for mutual exclusion (init = 1)
sem_t empty;   // counting semaphore: free  slots        (init = N)
sem_t full;    // counting semaphore: filled slots        (init = 0)

// ─── Shared counters (protected by mutex) ────────────────────────────────────
int produced_count = 0;
int consumed_count = 0;

// ════════════════════════════════════════════════════════════════════════════
//  Helper functions
// ════════════════════════════════════════════════════════════════════════════

int produce_item(int producer_id) {
    int item = rand() % 100 + 1;
    printf("  [Producer-%d] Produced  item = %d\n", producer_id, item);
    sleep(1);   // simulate time to produce
    return item;
}

void insert_item(int item) {
    buffer[in] = item;
    in = (in + 1) % N;
}

int remove_item() {
    int item = buffer[out];
    out = (out + 1) % N;
    return item;
}

void consume_item(int item, int consumer_id) {
    printf("  [Consumer-%d] Consumed  item = %d\n", consumer_id, item);
    sleep(1);   // simulate time to consume
}

// ════════════════════════════════════════════════════════════════════════════
//  Producer Thread
// ════════════════════════════════════════════════════════════════════════════
void *producer(void *arg) {
    int id = *((int *)arg);
    free(arg);

    while (1) {
        // ── decide whether this producer still has work ──
        sem_wait(&mutex);
        if (produced_count >= total_items) {
            sem_post(&mutex);
            break;
        }
        produced_count++;
        sem_post(&mutex);

        // ── produce outside critical section ──
        int item = produce_item(id);

        // ── SEMAPHORE LOGIC ──────────────────────────────
        sem_wait(&empty);   // wait() : wait if buffer is full
        sem_wait(&mutex);   // wait() : enter critical section

        insert_item(item);
        printf("  [Producer-%d] Inserted  item=%d  into slot=%d\n",
               id, item, (in - 1 + N) % N);

        sem_post(&mutex);   // signal() : leave critical section
        sem_post(&full);    // signal() : notify consumer: one more item ready
        // ─────────────────────────────────────────────────
    }

    printf("  [Producer-%d] Finished.\n", id);
    return NULL;
}

// ════════════════════════════════════════════════════════════════════════════
//  Consumer Thread
// ════════════════════════════════════════════════════════════════════════════
void *consumer(void *arg) {
    int id = *((int *)arg);
    free(arg);

    while (1) {
        // ── SEMAPHORE LOGIC ──────────────────────────────
        sem_wait(&full);    // wait() : wait if buffer is empty
        sem_wait(&mutex);   // wait() : enter critical section

        if (consumed_count >= total_items) {
            sem_post(&mutex);
            sem_post(&full);    // let other consumers also exit
            break;
        }
        consumed_count++;

        int item = remove_item();
        printf("  [Consumer-%d] Removed   item=%d  from slot=%d\n",
               id, item, (out - 1 + N) % N);

        sem_post(&mutex);   // signal() : leave critical section
        sem_post(&empty);   // signal() : notify producer: one slot is free
        // ─────────────────────────────────────────────────

        consume_item(item, id);
    }

    printf("  [Consumer-%d] Finished.\n", id);
    return NULL;
}

// ════════════════════════════════════════════════════════════════════════════
//  Main
// ════════════════════════════════════════════════════════════════════════════
int main() {
    int num_producers, num_consumers;

    printf("╔══════════════════════════════════════════════════╗\n");
    printf("║   Producer-Consumer Problem  (Semaphore in C)   ║\n");
    printf("╚══════════════════════════════════════════════════╝\n\n");

    printf("Enter buffer size           (N) : "); scanf("%d", &N);
    printf("Enter total items to produce    : "); scanf("%d", &total_items);
    printf("Enter number of producers       : "); scanf("%d", &num_producers);
    printf("Enter number of consumers       : "); scanf("%d", &num_consumers);

    // ── Allocate buffer ──────────────────────────────────────────────────────
    buffer = (int *)calloc(N, sizeof(int));

    // ── Initialize semaphores ────────────────────────────────────────────────
    sem_init(&mutex, 0, 1);   // mutex = 1
    sem_init(&empty, 0, N);   // empty = N  (all slots free)
    sem_init(&full,  0, 0);   // full  = 0  (no items yet)

    printf("\n─── Simulation Start ───────────────────────────────\n\n");

    // ── Create threads ───────────────────────────────────────────────────────
    pthread_t prod_threads[num_producers];
    pthread_t cons_threads[num_consumers];

    for (int i = 0; i < num_producers; i++) {
        int *id = malloc(sizeof(int)); *id = i + 1;
        pthread_create(&prod_threads[i], NULL, producer, id);
    }
    for (int i = 0; i < num_consumers; i++) {
        int *id = malloc(sizeof(int)); *id = i + 1;
        pthread_create(&cons_threads[i], NULL, consumer, id);
    }

    // ── Wait for all threads ─────────────────────────────────────────────────
    for (int i = 0; i < num_producers; i++)
        pthread_join(prod_threads[i], NULL);

    for (int i = 0; i < num_consumers; i++)
        pthread_join(cons_threads[i], NULL);

    // ── Cleanup ──────────────────────────────────────────────────────────────
    sem_destroy(&mutex);
    sem_destroy(&empty);
    sem_destroy(&full);
    free(buffer);

    printf("\n─── Simulation End ─────────────────────────────────\n");
    printf("  Total produced : %d\n", produced_count);
    printf("  Total consumed : %d\n", consumed_count);
    printf("╔══════════════════════════════════════════════════╗\n");
    printf("║                    All Done!                    ║\n");
    printf("╚══════════════════════════════════════════════════╝\n");

    return 0;
}
